File: 2021-03-ACCU/ConceptsTraitsTypeClassesProtocols/add.swift

// func add(_ a: Int, _ b: Int) -> Int { a + b }
func add<T: Numeric>(_ a: T, _ b: T) -> T { a + b }
//func add<T>(_ a: T, _ b: T) -> T { a + b } // fails

print(add(1, 2))

File: Talks/2020-11-MeetingC++/ConceptsTraitsTypeClassesProtocols/add.rs

// https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0a3c8942297fd09aba64f869dcbebf62

// fn add(a: i32, b: i32) -> i32 { a + b }
fn add<T: std::ops::Add<Output = T>>(a: T, b: T) -> T { a + b }
// fn add<T>(a: T, b: T) -> T { a + b } // fails

fn main() {
    println!("{}\n", add(1, 2));
}
